channels/channelmapper.go

```diff
@@ -60,79 +113,149 @@ func GetDefaultSyncFunction(scopeName, collectionName string) string {
 
 }
 
-func (mapper *ChannelMapper) MapToChannelsAndAccess(body map[string]interface{}, oldBodyJSON string, metaMap map[string]interface{}, userCtx map[string]interface{}) (*ChannelMapperOutput, error) {
-	numberFixBody := ConvertJSONNumbers(body)
-	numberFixMetaMap := ConvertJSONNumbers(metaMap)
+func (mapper *ChannelMapper) Function() string {
+	return mapper.fnSource
+}
+
+// This function is DEPRECATED. It's currently used in some tests that can't easily be changed.
+// Its current implementation is a kludge, and it shouldn't be used in production.
+func (mapper *ChannelMapper) SetFunction(fnSource string) error {
```

Comment:
It might only get called under tests that run under GSI/CBS, but we do call it here https://github.com/couchbase/sync_gateway/blob/3e36d801b8c6a659622d3d148dfcedd772c0b5ca/db/database.go#L511 to determine if we need to do a resync.

Reply:
That call is also in NewDatabaseContext, so it's only initializing the sync fn, not changing it, and SetFunction is not called. `dbCollection` was only instantiated 5 lines earlier, and newDatabaseCollection doesn't initialize c.ChannelMapper, so it's still nil.

The _only_ case where ChannelMapper.SetFunction is called is when a unit test calls UpdateSyncFun directly on an already-open database. That seems like a bad thing for a test to do since it's an otherwise-impossible change to the database.